spec/unit/matrixrtc/MembershipManager.spec.ts

```diff
@@ -0,0 +1,611 @@
+/**
+ * @jest-environment ./spec/unit/matrixrtc/testEnvironment.ts
+ */
+/*
+Copyright 2025 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { type Mock } from "jest-mock";
+import waitForExpect from "wait-for-expect";
+
+import { EventType, HTTPError, MatrixError, type Room } from "../../../src";
+import { type Focus, type LivekitFocusActive, type SessionMembershipData } from "../../../src/matrixrtc";
+import { LegacyMembershipManager } from "../../../src/matrixrtc/MembershipManager";
+import { makeMockClient, makeMockRoom, membershipTemplate, mockCallMembership, type MockClient } from "./mocks";
+import { flushPromises } from "../../test-utils/flushPromises";
+// import { MembershipManager } from "../../../src/matrixrtc/NewMembershipManager";
+
+function waitForMockCall(method: any, returnVal?: any) {
+    return new Promise<void>((resolve) => {
+        (method as Mock).mockImplementation(() => {
+            resolve();
+            return returnVal;
+        });
+    });
+}
+
+function createAsyncHandle(method: any) {
+    const handle: { resolve?: (...args: unknown[]) => void; reject?: (...args: any[]) => void } = {};
+    (method as Mock).mockImplementation(() => {
+        return new Promise((resolve, reject) => {
+            handle.reject = reject;
+            handle.resolve = resolve;
+        });
+    });
```

Comment:
utils/defer might also save you a line or two here. Also maybe the name should involve the word 'defer' as that's essentially what this is and I got a bit confused by 'handle'.